### ISSUE TITLE
On saving/loading checkpoints

### ISSUE THREAD

**Issue by axelurbiz** [statsmaths/kerasR#6]
Hi,

This is more of a question than an issue. 

I’d like to use checkpoints to save the model weights to the same file, only if the validation accuracy improves. I’ve put together a reproducible example below on my attempt for doing this, but I don't think this is done right. In particular, the output from the last step shows that my validation accuracy is not improving at any interaction. However, if I fit the same model without checkpoints (using the line of code commented out), validation accuracy improves with almost every single iteration. 

```r
### Fitting a model with checkpoints

library(reticulate)
use_python("/anaconda/bin/python")
library(kerasR)

db <- MASS::Boston
X <- as.matrix(db[, 1:13])
Y <- as.matrix(db[, 14])

model <- Sequential()
model$add(Dense(units = 3, input_shape = 13, kernel_initializer = "normal"))
model$add(Activation("relu"))
model$add(Dense(units = 1, kernel_initializer = "normal"))
keras_compile(model, loss = "mean_squared_error", optimizer = Adam())

filepath <- "/Users/lguelman/Library/Mobile Documents/com~apple~CloudDocs/LG_Files/Programming/TensorFlow/keras/rweights.ch14.hdf5"
checkpoint <- ModelCheckpoint(filepath, monitor = "val_loss", verbose=1, save_best_only=TRUE, mode= "max")
callbacks_list <- list(checkpoint)

#keras_fit(model, X, Y, batch_size = 5, epochs = 100, verbose = 1,
#          validation_split = 0.33) # no check points, works ok.

keras_fit(model, X, Y, batch_size = 5, epochs = 100, verbose = 0,
          validation_split = 0.33, callbacks = callbacks_list)

### Loading a saved model

model <- Sequential()
model$add(Dense(units = 3, input_shape = 13, kernel_initializer = "normal"))
model$add(Activation("relu"))
model$add(Dense(units = 1, kernel_initializer = "normal"))
model$load_weights("/Users/lguelman/Library/Mobile Documents/com~apple~CloudDocs/LG_Files/Programming/TensorFlow/keras/rweights.ch14.hdf5")
keras_compile(model, loss = "mean_squared_error", optimizer = Adam())

db <- MASS::Boston
X <- as.matrix(db[, 1:13])
Y <- as.matrix(db[, 14])

model$evaluate(X, Y)[[1]]

```
...
Epoch 00001: val_loss did not improve
Epoch 00002: val_loss did not improve
...
Epoch 00099: val_loss did not improve

Note that I’m passing the checkpoint as a list in the call to `keras_fit`. Not sure if this is correct. 

Thank you.

Regards,
Axel.

**Comment by statsmaths** [comment]
@axelurbiz This is an interested code snippet. Here are a few thoughts:
- when I run the keras_fit call without the callbacks, while the model loss improves run over run the val_loss does not; it seems to go up and down and have its best value around Epoch 40.
- you want to set the "mode" in `ModelCheckpoint` to "min" or "auto. At the present you are trying to maximize the loss rather than minimize it
- if you are primarily an R programmer, remember that keras is an object-orient library. If you want to compare the two versions of keras_fit you need to re-construct the model object. Otherwise each fit is just picking up where the last one left off.
Let me know if that does not solve your problem. I think switching the mode of the model checkpoint object will fix most of the issues you had.

**Comment by axelurbiz** [comment]
Good catch! Switching the mode fixed the issue.

Thanks Taylor!
